A shell is a program that provides an interface between you (the user) and the operating system (like Windows, Linux, or macOS). Think of it like a translator that helps you communicate with your computer using commands.

There are different types of shells:

Command Prompt: The default shell for Windows.
PowerShell: An advanced shell with more capabilities, also for Windows.
Bash (Bourne Again Shell): A popular shell for Linux and macOS systems.
Git Bash is a shell environment for Windows that provides Bash (the Linux-style shell) and integrates it with some additional features from Git.
Bash on Windows: Normally, Windows uses Command Prompt or PowerShell, but Bash is a popular shell used on Linux systems. Git Bash brings the Bash shell to Windows, so you can use Linux-like commands (ls, pwd, rm, etc.) on your Windows system.

A bash shell is like a command prompt but with commands similar to those used on Linux. Since we're using Git Bash and MinGW, these commands will behave like we're in a Linux-like environment on our Windows system. Git Bash is installed on Windows as a separate terminal program.
It uses MinGW (Minimalist GNU for Windows) as its backend, which provides the necessary tools and libraries to mimic a Linux environment.

%%bash
This command is specific to Jupyter notebooks and tells the notebook to run the following commands in a bash shell.
The %%bash is called a cell magic command. It allows you to run bash shell commands directly in a code cell. This means that any commands you write in that cell will be executed as if they were typed in a bash shell, instead of being interpreted as Python code.

Here are a few common commands you can use in Git Bash:

ls – Lists files and directories in the current directory.
pwd – Prints the current directory path.
cd <directory> – Changes to a different directory.
touch <filename> – Creates an empty file.
rm <filename> – Deletes a file.
git status – Shows the current status of your Git repository.
git clone <url> – Clones a Git repository from a URL.

Why Not Just Use Python Code?
Using bash commands is simpler and more efficient than using Python code, especially when interacting with the operating system. For example:
File System Operations: ls, pwd, cd, rm are simpler than their Python equivalents (os.listdir(), os.getcwd(), os.remove()).
Version Control: Managing Git commands is easier using shell commands (git status, git commit, etc.) rather than using Python libraries like subprocess.
Package Management: Installing packages with pip or conda is easier in bash than using subprocess or Python methods.

In [1]:
%%bash
pwd

#Present working directory

/g/My Drive/Upskilling/Github repository/Rep1/Linux_shell


In [2]:
%%bash
ls

#List files in current folder

Linux_shell_course.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb


In [8]:
%%bash
ps

#Which shell is responding to our commands- bash shell 

      PID    PPID    PGID     WINPID   TTY         UID    STIME COMMAND
      148     147     147       1996  ?         197609 09:57:43 /usr/bin/ps
      147       1     147       2756  ?         197609 09:57:43 /usr/bin/bash


The ps command shows the current running processes in your bash shell.

The first line shows that bash is running with a process ID (PID).
The second line shows the ps command itself running with a PID.
This tells us what processes are running within your bash shell and gives information like when they started (STIME) and what command is being run (COMMAND).

In [6]:
%%bash
uname

#name of operating system


MINGW64_NT-10.0-22631


The uname command outputs the name of your operating system or kernel.

Kernel (in an operating system context): A core part of the operating system that manages hardware resources and coordinates between hardware and software.
Kernel (in a Jupyter notebook context): An execution environment that runs your code, keeps track of its state, and returns the results.

MinGW (Minimalist GNU for Windows), provides a Unix-like environment.
This output indicates that you are using MinGW64 on a Windows system (NT-10.0-22631 refers to your Windows version).
It means your bash shell is is built on top of MinGW.

Since we're using Git Bash and MinGW, these commands will behave like we're in a Linux-like environment on our Windows system



The Jupyter/VS Code kernel and the operating system (OS) kernel are different concepts:

Operating System Kernel:
The core component of an OS (like Windows, Linux, or macOS).
Manages hardware resources, processes, memory, and device communication.
Acts as the intermediary between hardware and software.

Jupyter/VS Code Kernel:
A specific execution environment that runs code in a programming language (e.g., Python, R, Julia). Jupyter Kernels:
A Jupyter kernel is an execution environment that determines how the code in your notebook is run.
Kernels can be for any language (Python, Bash, R, Julia, etc.).
Each kernel is linked to a specific interpreter or environment.


Conda Environment:
A Conda environment is an isolated directory that contains its own Python interpreter, libraries, and packages.
It helps you manage dependencies and avoid conflicts between different projects.
When you create a Conda environment named simulations_env, it has its own Python interpreter and packages.

You can then register an environment environment as a Jupyter kernel. When you select this kernel, the code you run will use the packages and interpreter from sim# 1. List all available kernels
jupyter kernelspec list

# 2. Start Jupyter notebook with a specific kernel (replace python3 with name)
jupyter notebook --NotebookApp.default_kernel_name=python3

# 3. Verify the active kernel within a notebook cell
# (Run the following in the Jupyter notebook itself- its a python code)
import sys
print(sys.executable)

# 4. Install a new kernel based on a Conda environment
python -m ipykernel install --user --name simulations_env --display-name "Simulations Environment"

# 5. Remove an existing kernel (replace bash with name of kernel)
jupyter kernelspec uninstall bash

# 6. conda info --envs
list conda environments
ulations_env.

In [2]:
%%bash
jupyter kernelspec list

0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.


Available kernels:
  python3    C:\Users\kavis\.conda\envs\simulations_env\Lib\site-packages\ipykernel\resources


In [1]:
!conda info --envs


# conda environments:
#
base                     C:\ProgramData\anaconda3
mayavi_env               C:\Users\kavis\.conda\envs\mayavi_env
simulations_env       *  C:\Users\kavis\.conda\envs\simulations_env



The base environment is created during Anaconda’s installation, and by default, it goes to a shared location on your system (e.g., C:\ProgramData\anaconda3).
Any new environment you create using Conda goes under a hidden .conda folder in your user directory (C:\Users\kavis\.conda\envs\), unless you specify a different location.
This separation helps keep the default setup and custom environments independent from each other.

In [7]:
%%bash
clear # used to clear the screen
# or use ctrl+L

In [9]:
%%bash
exit
# or ctrl+D to come out of shell

In [10]:
%%bash
ls -a
# -a option gives hidden files also

.
..
Linux_shell_course.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb


In [11]:
%%bash
ls -l
# -l option gives long format

total 270
-rw-r--r-- 0 kavis 197609  11069 Oct  6 11:21 Linux_shell_course.ipynb
-rw-r--r-- 0 kavis 197609   8173 Oct  5 19:51 SC_Week0_Unit0.ipynb
-rw-r--r-- 0 kavis 197609 230730 Oct  5 19:51 SC_Week0_Unit1.ipynb
-rw-r--r-- 0 kavis 197609  16186 Oct  5 20:28 SC_Week0_Unit2.ipynb
-rw-r--r-- 0 kavis 197609   3531 Oct  5 19:52 SC_Week0_Unit4.ipynb
-rw-r--r-- 0 kavis 197609   4159 Oct  5 19:51 SC_Week0_Unit5.ipynb


In [12]:
%%bash
ls --help
#summary of how to use the command

Usage: ls [OPTION]... [FILE]...
List information about the FILEs (the current directory by default).
Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.

Mandatory arguments to long options are mandatory for short options too.
  -a, --all                  do not ignore entries starting with .
  -A, --almost-all           do not list implied . and ..
      --author               with -l, print the author of each file
  -b, --escape               print C-style escapes for nongraphic characters
      --block-size=SIZE      with -l, scale sizes by SIZE when printing them;
                               e.g., '--block-size=M'; see SIZE format below
  -B, --ignore-backups       do not list implied entries ending with ~
  -c                         with -lt: sort by, and show, ctime (time of last
                               modification of file status information);
                               with -l: show ctime and sort by name;
                               othe

In [15]:
%%bash
man ls
clear
# doesn't work with vs code
# manual page for the command, which provides detailed documentation.

/usr/bin/bash: line 1: man: command not found


Filesystem Hierarchy Standard followed by all linux systems
FHS 3.0 released on June 03, 2015
Available at
https://refspecs.linuxfoundation.org/fhs.shtml

In [28]:
%%bash
cd ..
##change directory to parent directory of current directory (parent directory means just one level up)
pwd

cd . 
## single dot stands for present directory so, no change
# cd . means stay in the current directory.
pwd

cd "g/My Drive/Upskilling/Github repository"  
pwd
#this path is relative to current directory and it doesn't actually exist

cd "/g/My Drive/Upskilling/Github repository"
pwd
#absolute path

# use the command cd ~ or cd $HOME
cd ~
# The home directory is the starting directory for a user
pwd


/g/My Drive/Upskilling/Github repository/Rep1
/g/My Drive/Upskilling/Github repository/Rep1


/usr/bin/bash: line 10: cd: g/My Drive/Upskilling/Github repository: No such file or directory


/g/My Drive/Upskilling/Github repository/Rep1
/g/My Drive/Upskilling/Github repository
/c/Users/kavis


In [35]:
%%bash
pwd

cd     # cd without arguement takes to the home directory
pwd

cd /    #root directory  (/) is the highest level in the filesystem hierarchy 
pwd
ls -a

#When you list all items in a directory using ls -a, you will see ..:
#   . represents the current directory.
#  .. always represents the parent directory, even if it’s a reference to the same directory (as is the case in the root directory).

cd /
pwd
cd ..
pwd

/g/My Drive/Upskilling/Github repository/Rep1/Linux_shell
/c/Users/kavis
/
.
..
LICENSE.txt
ReleaseNotes.html
bin
cmd
dev
etc
git-bash.exe
git-cmd.exe
mingw64
proc
tmp
unins001.dat
unins001.exe
unins001.msg
usr
/
/


In [37]:
%%bash
pwd
cd /c
ls -l

/g/My Drive/Upskilling/Github repository/Rep1/Linux_shell
total 10793673
drwxr-xr-x 1 kavis 197609          0 Oct 20  2022 $Recycle.Bin
-rw-r--r-- 1 kavis 197609         40 Oct 21  2022 2CA0AE717CAA
lrwxrwxrwx 1 kavis 197609          8 Oct 11  2022 Documents and Settings -> /c/Users
drwxr-xr-x 1 kavis 197609          0 Jul 30  2022 Drivers
-rw-r--r-- 1 kavis 197609      12288 Jul 12 09:14 DumpStack.log
-rw-r--r-- 1 kavis 197609      12288 Oct  5 18:18 DumpStack.log.tmp
drwxr-xr-x 1 kavis 197609          0 May  4 13:43 OneDriveTemp
drwxr-xr-x 1 kavis 197609          0 May  7  2022 PerfLogs
drwxr-xr-x 1 kavis 197609          0 Oct  6 09:13 Program Files
drwxr-xr-x 1 kavis 197609          0 Jun 11 08:47 Program Files (x86)
drwxr-xr-x 1 kavis 197609          0 Sep 19 22:02 ProgramData
drwxr-xr-x 1 kavis 197609          0 Aug 10  2023 Recovery
-rw-r--r-- 1 kavis 197609       1505 Dec 19  2023 Setup.log
drwxr-xr-x 1 kavis 197609          0 Oct  6 09:12 System Volume Information
drwxr-xr-x 1 

#### Structure of Each Line:
1. **Permissions**: (`drwxr-xr-x`)
   - This shows the file permissions for the owner, group, and others.
   - first letter -  `d` indicates a **directory** (if it were a file, it would be `-`).
                     'l' indicates symbolic link (shortcut)
                     '-' indicates file
                     'c' character device
                     'b' block devices (hard discs)
   - next three for owner - `rwx` means **read, write, and execute** permissions for the owner.
   - next three for group - `r-x` means **read and execute** permissions for the group.
   - next three for others- `r-x` means **read and execute** permissions for others.

2. **Links**: (`1`)
   - The number of links or directories inside this directory.

3. **Owner**: (`kavis`)
   - The username of the owner of the file or directory.

4. **Group**: (`197609`)
   - The group ID associated with the file.

5. **Size**: (`0`, `40`, `12288`, etc.)
   - The size of the file in bytes. For directories, it’s usually `0` since it doesn’t directly reflect the size of contents.

6. **Timestamp**: (`Oct 6 14:21`)
   - The date and time when the file or directory was last modified.

7. **File/Directory Name**: (`$Recycle.Bin`, `Documents and Settings`, etc.)
   - The name of the file or directory.

#### Listing Explanation:

**`Documents and Settings -> /c/Users`**:
   - This is a **symbolic link** (shortcut) that points to `/c/Users`. It means that trying to access `Documents and Settings` will actually redirect you to the `/c/Users` directory.

**`Drivers`**:
   - Likely a folder containing drivers for the system. Drivers are software that help the operating system communicate with hardware devices.

**`Program Files`** and **`Program Files (x86)`**:
   - Standard directories in Windows for installing 64-bit and 32-bit applications, respectively.

**`ProgramData`**:
   - A directory used to store application data that is not specific to a user. It’s shared among all users of the system.

**`Users`**:
    - This is where user profiles are stored on Windows, such as `Desktop`, `Documents`, `Downloads`, and so on for each user.

**`Windows`**:
    - The main directory containing the Windows operating system files.

**`hiberfil.sys`**:
    - A **system file** used by Windows to store the system state when it goes into **hibernation**. This file is often large and contains everything stored in memory (RAM).

In [5]:
%%bash
pwd
cd /
ls -l

cd -  # go to previous directory

/g/My Drive/Upskilling/Github repository/Rep1/Linux_shell
total 5520
-rw-r--r-- 1 kavis 197609   18765 Sep 24 12:26 LICENSE.txt
-rw-r--r-- 1 kavis 197609  268097 Sep 24 12:26 ReleaseNotes.html
drwxr-xr-x 1 kavis 197609       0 Oct  5 17:34 bin
drwxr-xr-x 1 kavis 197609       0 Oct  5 17:34 cmd
drwxr-xr-x 1 kavis 197609       0 Oct  5 17:34 dev
drwxr-xr-x 1 kavis 197609       0 Oct  5 17:34 etc
-rwxr-xr-x 1 kavis 197609  139144 Sep 24 12:05 git-bash.exe
-rwxr-xr-x 1 kavis 197609  138616 Sep 24 12:05 git-cmd.exe
drwxr-xr-x 1 kavis 197609       0 Oct  5 17:34 mingw64
dr-xr-xr-x 8 kavis 197609       0 Oct  6 16:00 proc
drwxr-xr-x 1 kavis 197609       0 Oct  6 15:59 tmp
-rw-r--r-- 1 kavis 197609 1315870 Oct  5 17:34 unins001.dat
-rwxr-xr-x 1 kavis 197609 3384040 Oct  5 17:34 unins001.exe
-rw-r--r-- 1 kavis 197609   24183 Oct  5 17:34 unins001.msg
drwxr-xr-x 1 kavis 197609       0 Oct  5 17:34 usr
/g/My Drive/Upskilling/Github repository/Rep1/Linux_shell


/bin -       essential command binaries - files in machine language for processor
/usr/bin -   User commands
/usr/lib -   Libraries

In [25]:
%%bash
date
date -R

cal

Sun Oct  6 16:25:35 IST 2024
Sun, 06 Oct 2024 16:25:35 +0530


/usr/bin/bash: line 4: cal: command not found


CalledProcessError: Command 'b'date\ndate -R\n\ncal\n'' returned non-zero exit status 127.